File: android/wear/src/main/kotlin/app/dozealert/wear/TripState.kt
package app.dozealert.wear

data class TripState(
    val state: String = "idle",
    val destinationName: String = "",
    val distanceKm: Double = 0.0,
    val distanceReady: Boolean = false,
    val stopsRemaining: Int = -1,
    val transitActive: Boolean = false,
    val lineLabel: String = "",
    val tripConcern: String = "",
    val gpsStale: Boolean = false,
    val directionLabel: String = "",
    val nextStopName: String = "",
    val wakeStopCount: Int = -1,
    val alarmActive: Boolean = false,
    val hasDestination: Boolean = false,
    val alarmStopName: String = "",
    val alarmHeadline: String = "",
    val alarmUiHeadline: String = "",
    val alarmSubline: String = "",
    val updatedAt: Long = 0L,
) {
    val isMonitoring: Boolean
        get() = state == "monitoring"

    val hasTripConcern: Boolean
        get() = tripConcern.isNotBlank()

    val showTripConcernBanner: Boolean
        get() = isMonitoring && hasTripConcern && !alarmActive

    val canStart: Boolean
        get() = hasDestination && state == "idle"

    val canStop: Boolean
        get() = isMonitoring || state == "arrived" || alarmActive

    enum class StatusKind {
        Idle,
        Ready,
        Monitoring,
        Alarm,
        Arrived,
        Missed,
    }

    val statusKind: StatusKind
        get() = when {
            alarmActive -> StatusKind.Alarm
            state == "arrived" -> StatusKind.Arrived
            state == "missed" -> StatusKind.Missed
            isMonitoring -> StatusKind.Monitoring
            hasDestination -> StatusKind.Ready
            else -> StatusKind.Idle
        }

    val headline: String
        get() = when {
            alarmActive -> alarmUiHeadline.ifBlank { "GET READY" }
            !hasDestination -> "Set up on phone"
            isMonitoring && hasTripConcern && tripConcern == "wrong_direction" ->
                "Wrong direction?"
            isMonitoring && hasTripConcern && tripConcern == "unlikely_route" ->
                "Route uncertain"
            isMonitoring && gpsStale -> "GPS signal weak"
            isMonitoring && transitActive && stopsRemaining >= 0 -> when (stopsRemaining) {
                0 -> "At destination stop"
                1 -> "1 stop to go"
                else -> "$stopsRemaining stops to go"
            }
            isMonitoring && distanceReady -> String.format("%.1f km left", distanceKm)
            isMonitoring -> "Watching your trip"
            state == "arrived" -> "You've arrived"
            state == "missed" -> "Trip missed"
            hasDestination -> destinationName.ifBlank { lineLabel }.ifBlank { "Your trip" }
            else -> "Set up on phone"
        }

    val alarmPrimaryStopName: String
        get() = alarmStopName.ifBlank { destinationName }

    val alarmContextLine: String
        get() {
            val synced = alarmSubline.trim()
            if (synced.isNotEmpty()) {
                return synced
            }

            if (wakeStopCount > 0 && stopsRemaining >= 0) {
                val capped = minOf(stopsRemaining, wakeStopCount)
                return when (capped) {
                    0 -> "Time to get off"
                    1 -> "1 more stop to go"
                    else -> "$capped more stops to go"
                }
            }

            return if (destinationName.isNotBlank()) {
                "Within wake radius"
            } else {
                "Within wake radius"
            }
        }

    val tripConcernDetail: String
        get() = when (tripConcern) {
            "wrong_direction" -> "Check line & direction on phone."
            "unlikely_route" -> "Confirm your trip on phone."
            else -> "Check your line and direction on phone."
        }

    val subline: String
        get() = when {
            alarmActive -> alarmContextLine
            !hasDestination -> "Pick a destination on phone"
            isMonitoring && transitActive -> buildMonitoringSubline()
            isMonitoring && !distanceReady -> destinationName.ifBlank { "Waiting for GPS…" }
            isMonitoring -> destinationName.ifBlank { "Trip in progress" }
            hasDestination && !isMonitoring && state != "arrived" && state != "missed" -> {
                val line = lineLabel.ifBlank { null }
                when {
                    line != null -> line
                    else -> "Tap Start when on board"
                }
            }
            hasDestination -> {
                val dest = destinationName.ifBlank { null }
                val line = lineLabel.ifBlank { null }
                when {
                    dest != null && line != null -> "$dest · $line"
                    dest != null -> dest
                    line != null -> line
                    else -> "Tap Start when you are on board"
                }
            }
            else -> "Open phone to set destination"
        }

    private fun buildMonitoringSubline(): String {
        val parts = mutableListOf<String>()
        if (nextStopName.isNotBlank()) {
            parts.add("Next: $nextStopName")
        }
        val line = lineLabel.ifBlank { null }
        val dest = destinationName.ifBlank { null }
        when {
            line != null && dest != null -> parts.add("$line · $dest")
            line != null -> parts.add(line)
            dest != null -> parts.add(dest)
        }
        if (directionLabel.isNotBlank() && !hasTripConcern) {
            parts.add(directionLabel)
        }
        if (gpsStale) {
            parts.add("Last known position")
        }
        if (hasTripConcern) {
            parts.add("Check line on phone")
        }
        return parts.joinToString(" · ").ifBlank { "Waiting for route data…" }
    }

    val statusLabel: String
        get() = when (statusKind) {
            StatusKind.Alarm -> "Alarm"
            StatusKind.Arrived -> "Arrived"
            StatusKind.Missed -> "Missed"
            StatusKind.Monitoring -> "Watching"
            StatusKind.Ready -> "Ready"
            StatusKind.Idle -> "Idle"
        }

    val tileLine: String
        get() = when {
            alarmActive -> alarmPrimaryStopName.ifBlank { alarmUiHeadline }.ifBlank { "GET READY" }
            isMonitoring && hasTripConcern -> "Check line on phone"
            isMonitoring && gpsStale -> "GPS weak"
            isMonitoring && transitActive && stopsRemaining >= 0 -> when (stopsRemaining) {
                0 -> "At stop"
                1 -> "1 stop to go"
                else -> "$stopsRemaining stops to go"
            }
            isMonitoring && nextStopName.isNotBlank() -> "Next: $nextStopName"
            isMonitoring && distanceReady -> String.format("%.1f km", distanceKm)
            isMonitoring -> "Watching"
            hasDestination -> destinationName
                .ifBlank { if (transitActive) lineLabel else "" }
                .ifBlank { "DozeAlert" }
            else -> "DozeAlert"
        }

    companion object {
        fun fromDataMap(map: com.google.android.gms.wearable.DataMap): TripState {
            return TripState(
                state = map.getString("state", "idle"),
                destinationName = map.getString("destinationName", ""),
                distanceKm = map.getDouble("distanceKm", 0.0),
                distanceReady = map.getBoolean("distanceReady", false),
                stopsRemaining = map.getInt("stopsRemaining", -1),
                transitActive = map.getBoolean("transitActive", false),
                lineLabel = map.getString("lineLabel", ""),
                tripConcern = map.getString("tripConcern", ""),
                gpsStale = map.getBoolean("gpsStale", false),
                directionLabel = map.getString("directionLabel", ""),
                nextStopName = map.getString("nextStopName", ""),
                wakeStopCount = map.getInt("wakeStopCount", -1),
                alarmActive = map.getBoolean("alarmActive", false),
                hasDestination = map.getBoolean("hasDestination", false),
                alarmStopName = map.getString("alarmStopName", ""),
                alarmHeadline = map.getString("alarmHeadline", ""),
                alarmUiHeadline = map.getString("alarmUiHeadline", ""),
                alarmSubline = map.getString("alarmSubline", ""),
                updatedAt = map.getLong("updatedAt", 0L),
            )
        }

        fun fromPreferences(prefs: android.content.SharedPreferences): TripState {
            return TripState(
                state = prefs.getString("state", "idle") ?: "idle",
                destinationName = prefs.getString("destinationName", "") ?: "",
                distanceKm = prefs.getFloat("distanceKm", 0f).toDouble(),
                distanceReady = prefs.getBoolean("distanceReady", false),
                stopsRemaining = prefs.getInt("stopsRemaining", -1),
                transitActive = prefs.getBoolean("transitActive", false),
                lineLabel = prefs.getString("lineLabel", "") ?: "",
                tripConcern = prefs.getString("tripConcern", "") ?: "",
                gpsStale = prefs.getBoolean("gpsStale", false),
                directionLabel = prefs.getString("directionLabel", "") ?: "",
                nextStopName = prefs.getString("nextStopName", "") ?: "",
                wakeStopCount = prefs.getInt("wakeStopCount", -1),
                alarmActive = prefs.getBoolean("alarmActive", false),
                hasDestination = prefs.getBoolean("hasDestination", false),
                alarmStopName = prefs.getString("alarmStopName", "") ?: "",
                alarmHeadline = prefs.getString("alarmHeadline", "") ?: "",
                alarmUiHeadline = prefs.getString("alarmUiHeadline", "") ?: "",
                alarmSubline = prefs.getString("alarmSubline", "") ?: "",
                updatedAt = prefs.getLong("updatedAt", 0L),
            )
        }
    }
}

fun TripState.persist(prefs: android.content.SharedPreferences) {
    prefs.edit()
        .putString("state", state)
        .putString("destinationName", destinationName)
        .putFloat("distanceKm", distanceKm.toFloat())
        .putBoolean("distanceReady", distanceReady)
        .putInt("stopsRemaining", stopsRemaining)
        .putBoolean("transitActive", transitActive)
        .putString("lineLabel", lineLabel)
        .putString("tripConcern", tripConcern)
        .putBoolean("gpsStale", gpsStale)
        .putString("directionLabel", directionLabel)
        .putString("nextStopName", nextStopName)
        .putInt("wakeStopCount", wakeStopCount)
        .putBoolean("alarmActive", alarmActive)
        .putBoolean("hasDestination", hasDestination)
        .putString("alarmStopName", alarmStopName)
        .putString("alarmHeadline", alarmHeadline)
        .putString("alarmUiHeadline", alarmUiHeadline)
        .putString("alarmSubline", alarmSubline)
        .putLong("updatedAt", updatedAt)
        .apply()
}

File: ios/DozeAlertWatchShared/TripState.swift
import Foundation

/// Shared trip-state model for Watch app + complications (mirrors Wear TripState).
struct TripState: Equatable, Codable {
  var state: String = "idle"
  var destinationName: String = ""
  var distanceKm: Double = 0
  var distanceReady: Bool = false
  var stopsRemaining: Int = -1
  var transitActive: Bool = false
  var lineLabel: String = ""
  var tripConcern: String = ""
  var gpsStale: Bool = false
  var directionLabel: String = ""
  var nextStopName: String = ""
  var wakeStopCount: Int = -1
  var alarmActive: Bool = false
  var hasDestination: Bool = false
  var alarmStopName: String = ""
  var alarmHeadline: String = ""
  var alarmUiHeadline: String = ""
  var alarmSubline: String = ""
  var updatedAt: Int64 = 0

  static let appGroupId = "group.app.dozealert"
  static let defaultsKey = "dozealert.watch.tripState"

  var isMonitoring: Bool { state == "monitoring" }

  var hasTripConcern: Bool { !tripConcern.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

  var showTripConcernBanner: Bool { isMonitoring && hasTripConcern && !alarmActive }

  var canStart: Bool { hasDestination && state == "idle" }

  var canStop: Bool { isMonitoring || state == "arrived" || alarmActive }

  enum StatusKind {
    case idle, ready, monitoring, alarm, arrived, missed
  }

  var statusKind: StatusKind {
    if alarmActive { return .alarm }
    if state == "arrived" { return .arrived }
    if state == "missed" { return .missed }
    if isMonitoring { return .monitoring }
    if hasDestination { return .ready }
    return .idle
  }

  var statusLabel: String {
    switch statusKind {
    case .alarm: return "Alarm"
    case .arrived: return "Arrived"
    case .missed: return "Missed"
    case .monitoring: return "Watching"
    case .ready: return "Ready"
    case .idle: return "Idle"
    }
  }

  var headline: String {
    if alarmActive {
      return alarmUiHeadline.isEmpty ? "GET READY" : alarmUiHeadline
    }
    if !hasDestination { return "Set up on phone" }
    if isMonitoring && hasTripConcern && tripConcern == "wrong_direction" {
      return "Wrong direction?"
    }
    if isMonitoring && hasTripConcern { return "Trip check needed" }
    if isMonitoring && gpsStale { return "GPS signal weak" }
    if isMonitoring && transitActive && stopsRemaining >= 0 {
      switch stopsRemaining {
      case 0: return "At destination stop"
      case 1: return "1 stop to go"
      default: return "\(stopsRemaining) stops to go"
      }
    }
    if isMonitoring && distanceReady {
      return String(format: "%.1f km left", distanceKm)
    }
    if isMonitoring { return "Watching your trip" }
    if state == "arrived" { return "You've arrived" }
    if state == "missed" { return "Trip missed" }
    if hasDestination {
      if !destinationName.isEmpty { return destinationName }
      if !lineLabel.isEmpty { return lineLabel }
      return "Your trip"
    }
    return "Set up on phone"
  }

  var alarmPrimaryStopName: String {
    alarmStopName.isEmpty ? destinationName : alarmStopName
  }

  var alarmContextLine: String {
    let synced = alarmSubline.trimmingCharacters(in: .whitespacesAndNewlines)
    if !synced.isEmpty { return synced }
    if wakeStopCount > 0 && stopsRemaining >= 0 {
      let capped = min(stopsRemaining, wakeStopCount)
      switch capped {
      case 0: return "Time to get off"
      case 1: return "1 more stop to go"
      default: return "\(capped) more stops to go"
      }
    }
    return "Within alert distance"
  }

  var tripConcernDetail: String {
    tripConcern == "wrong_direction"
      ? "Check line & direction on phone."
      : "Check your line and direction on phone."
  }

  var subline: String {
    if alarmActive { return alarmContextLine }
    if !hasDestination { return "Pick a destination on phone" }
    if isMonitoring && transitActive { return buildMonitoringSubline() }
    if isMonitoring && !distanceReady {
      return destinationName.isEmpty ? "Waiting for GPS…" : destinationName
    }
    if isMonitoring {
      return destinationName.isEmpty ? "Trip in progress" : destinationName
    }
    if hasDestination && !isMonitoring && state != "arrived" && state != "missed" {
      if !lineLabel.isEmpty { return lineLabel }
      return "Tap Start when on board"
    }
    if hasDestination {
      if !destinationName.isEmpty && !lineLabel.isEmpty {
        return "\(destinationName) · \(lineLabel)"
      }
      if !destinationName.isEmpty { return destinationName }
      if !lineLabel.isEmpty { return lineLabel }
      return "Tap Start when you are on board"
    }
    return "Open phone to set destination"
  }

  private func buildMonitoringSubline() -> String {
    var parts: [String] = []
    if !nextStopName.isEmpty { parts.append("Next: \(nextStopName)") }
    if !lineLabel.isEmpty && !destinationName.isEmpty {
      parts.append("\(lineLabel) · \(destinationName)")
    } else if !lineLabel.isEmpty {
      parts.append(lineLabel)
    } else if !destinationName.isEmpty {
      parts.append(destinationName)
    }
    if !directionLabel.isEmpty && !hasTripConcern {
      parts.append(directionLabel)
    }
    if gpsStale { parts.append("Last known position") }
    if hasTripConcern { parts.append("Check line on phone") }
    return parts.isEmpty ? "Waiting for route data…" : parts.joined(separator: " · ")
  }

  var complicationShortText: String {
    if alarmActive { return "WAKE" }
    if state == "arrived" { return "ARR" }
    if state == "missed" { return "MISS" }
    if isMonitoring && transitActive && stopsRemaining >= 0 {
      switch stopsRemaining {
      case 0: return "At"
      case 1: return "1 stp"
      default: return "\(stopsRemaining) stp"
      }
    }
    if isMonitoring && distanceReady {
      return String(format: "%.1f km", distanceKm)
    }
    if isMonitoring { return "ON" }
    if hasDestination { return "RDY" }
    return "OFF"
  }

  var complicationContentDescription: String {
    if alarmActive { return "Alarm · \(alarmContextLine)" }
    if state == "arrived" { return "You've arrived" }
    if state == "missed" { return "Trip missed" }
    if isMonitoring && transitActive && stopsRemaining >= 0 {
      switch stopsRemaining {
      case 0: return "Watching · At destination stop"
      case 1: return "Watching · 1 stop to go"
      default: return "Watching · \(stopsRemaining) stops to go"
      }
    }
    if isMonitoring && distanceReady {
      return String(format: "Watching · %.1f km left", distanceKm)
    }
    if isMonitoring { return "Watching your trip" }
    if hasDestination {
      let name = destinationName.isEmpty ? "Tap Start on phone" : destinationName
      return "Ready · \(name)"
    }
    return "DozeAlert idle"
  }

  static func from(dictionary: [String: Any]) -> TripState {
    TripState(
      state: dictionary["state"] as? String ?? "idle",
      destinationName: dictionary["destinationName"] as? String ?? "",
      distanceKm: (dictionary["distanceKm"] as? NSNumber)?.doubleValue
        ?? (dictionary["distanceKm"] as? Double) ?? 0,
      distanceReady: dictionary["distanceReady"] as? Bool ?? false,
      stopsRemaining: (dictionary["stopsRemaining"] as? NSNumber)?.intValue
        ?? (dictionary["stopsRemaining"] as? Int) ?? -1,
      transitActive: dictionary["transitActive"] as? Bool ?? false,
      lineLabel: dictionary["lineLabel"] as? String ?? "",
      tripConcern: dictionary["tripConcern"] as? String ?? "",
      gpsStale: dictionary["gpsStale"] as? Bool ?? false,
      directionLabel: dictionary["directionLabel"] as? String ?? "",
      nextStopName: dictionary["nextStopName"] as? String ?? "",
      wakeStopCount: (dictionary["wakeStopCount"] as? NSNumber)?.intValue
        ?? (dictionary["wakeStopCount"] as? Int) ?? -1,
      alarmActive: dictionary["alarmActive"] as? Bool ?? false,
      hasDestination: dictionary["hasDestination"] as? Bool ?? false,
      alarmStopName: dictionary["alarmStopName"] as? String ?? "",
      alarmHeadline: dictionary["alarmHeadline"] as? String ?? "",
      alarmUiHeadline: dictionary["alarmUiHeadline"] as? String ?? "",
      alarmSubline: dictionary["alarmSubline"] as? String ?? "",
      updatedAt: (dictionary["updatedAt"] as? NSNumber)?.int64Value
        ?? (dictionary["updatedAt"] as? Int64) ?? 0
    )
  }

  func asDictionary() -> [String: Any] {
    [
      "state": state,
      "destinationName": destinationName,
      "distanceKm": distanceKm,
      "distanceReady": distanceReady,
      "stopsRemaining": stopsRemaining,
      "transitActive": transitActive,
      "lineLabel": lineLabel,
      "tripConcern": tripConcern,
      "gpsStale": gpsStale,
      "directionLabel": directionLabel,
      "nextStopName": nextStopName,
      "wakeStopCount": wakeStopCount,
      "alarmActive": alarmActive,
      "hasDestination": hasDestination,
      "alarmStopName": alarmStopName,
      "alarmHeadline": alarmHeadline,
      "alarmUiHeadline": alarmUiHeadline,
      "alarmSubline": alarmSubline,
      "updatedAt": updatedAt,
    ]
  }

  static func loadFromAppGroup() -> TripState {
    let defaults = UserDefaults(suiteName: appGroupId) ?? .standard
    guard let dict = defaults.dictionary(forKey: defaultsKey) else {
      return TripState()
    }
    return TripState.from(dictionary: dict)
  }

  func saveToAppGroup() {
    let defaults = UserDefaults(suiteName: Self.appGroupId) ?? .standard
    defaults.set(asDictionary(), forKey: Self.defaultsKey)
    defaults.synchronize()
  }
}
